src/render/tile_renderer.cpp

```diff
@@ -0,0 +1,648 @@
+
+    common_->mvpManager.UpdateViewTransform();
+}
+
+Position2<int> render::TileRenderer::GetTileDrawAmount() const noexcept {
+    const auto matrix = glm::vec4(1, -1, 1, 1) / common_->mvpManager.GetMvpMatrix();
+    return Position2{LossyCast<int>(matrix.x / LossyCast<double>(tileWidth) * 2) + 2,
+                     LossyCast<int>(matrix.y / LossyCast<double>(tileWidth) * 2) + 2};
+}
+
+void render::TileRenderer::PrepareChunkRow(TRenderBuffer& r_layer,
+                                           const game::World& world,
+                                           std::mutex& world_gen_mutex,
+                                           Position2<int> row_start,
+                                           const int chunk_span,
+                                           Position2<int> render_tile_offset) const noexcept {
+    auto [tex_ids, readable_chunks] = world.GetChunkTexCoordIds(row_start);
+
+    if (readable_chunks < chunk_span) {
+        std::lock_guard gen_guard{world_gen_mutex};
+
+        for (int x = readable_chunks; x < chunk_span; ++x) {
+            const auto chunk_x = x + row_start.x;
+            world.QueueChunkGeneration({chunk_x, row_start.y});
+        }
+    }
+
+    // If the leftmost chunk is not readable, try other ones in row
+    // avoids annoying black screens when moving left
+    if (readable_chunks == 0) {
+        int i = 1; // Not 0, since we already know current chunk has no readable chunks
+        while (readable_chunks == 0 && i < chunk_span) {
+            std::tie(tex_ids, readable_chunks) = world.GetChunkTexCoordIds({row_start.x + i, row_start.y});
+            render_tile_offset.x += game::Chunk::kChunkWidth;
+            ++i;
+        }
+        row_start.x += i - 1; // i gets increment extra time at end of while
+    }
+
+    // Despite being readable, the chunk may not be generated
+    auto tex_ids_2 = tex_ids; // Cannot move the original pointer
+    for (int i = 0; i < readable_chunks; ++i) {
+        if (tex_ids_2[0] == 0) { // First tile, bottom layer of chunk
+            std::lock_guard gen_guard{world_gen_mutex};
+            world.QueueChunkGeneration({row_start.x + i, row_start.y});
+        }
+        tex_ids_2 += game::Chunk::kChunkArea * game::kTileLayerCount;
+    }
+
+    // Allocate for the maximum possible tile layers to render
+    const auto required_r_layer_capacity =
+        SafeCast<uint32_t>(chunk_span * game::Chunk::kChunkArea * game::kTileLayerCount);
+    if (required_r_layer_capacity >= r_layer.Capacity()) {
+        r_layer.Reserve(required_r_layer_capacity);
+        return;
+    }
+
+    // Prevents PrepareChunk from having a pixel x/y < 0
+    auto skip_tiles_top = -render_tile_offset.y / SafeCast<int>(tileWidth);
+    if (skip_tiles_top < 0) { // No tiles to skip
+        skip_tiles_top = 0;
+    }
+
+    auto tiles_prepare_y = (windowHeight_ - render_tile_offset.y) / SafeCast<int>(tileWidth);
```

Comment:
A subtraction between pixels (windowHeight) and tiles (render_tile_offset.y), this is incorrect.
Document this is tiles needed to reach the windowHeight